src/hexkit/providers/akafka/provider/eventsub.py

```diff
@@ -262,3 +429,154 @@ async def run(self, forever: bool = True) -> None:
+                event_value.decode("ascii")
+            ),
+        )
+
+        try:
+            await consumer.start()
+            yield cls(
+                consumer=consumer,
+                publisher=publisher,
+                dlq_topic=config.kafka_dlq_topic,
+                retry_topic=config.kafka_retry_topic,
+            )
+        finally:
+            await consumer.stop()
+
+    def __init__(
+        self,
+        *,
+        consumer: KafkaConsumerCompatible,
+        publisher: EventPublisherProtocol,
+        dlq_topic: str,
+        retry_topic: str,
+    ):
+        """Please do not call directly! Should be called by the `construct` method.
+
+        Args:
+        - `consumer`:
+            hands over a started AIOKafkaConsumer.
+        - `publisher`:
+            running instance of publishing provider that implements the
+            EventPublisherProtocol, such as KafkaEventPublisher.
+        - `dlq_topic`:
+            The name of the topic used to store failed events, to which the
+            KafkaDLQSubscriber subscribes.
+        - `retry_topic`:
+            The name of the topic used to requeue failed events.
+        """
+        self._consumer = consumer
+        self._publisher = publisher
+        self._dlq_topic = dlq_topic
+        self._retry_topic = retry_topic
+
+    async def _publish_to_retry(self, event: ConsumerEvent):
+        """Publish the event to the retry topic.
+
+        Events that lack a type or correlation_id in their headers are ignored.
+
+        Raises:
+        - `OriginalTopicError`:
+            Raised when the original topic is missing from the event.
+        """
+        event_label = get_event_label(event)
+        headers = headers_as_dict(event)
+
+        try:
+            type_ = headers["type"]
+            correlation_id = headers["correlation_id"]
+        except KeyError as err:
+            logging.warning("Ignored an event: %s. %s", event_label, err.args[0])
+            # acknowledge event receipt
+            await self._consumer.commit()
+            return
+
+        # Raise an error if the original topic is missing, because that is crucial
+        # information for publishing to the retry topic
+        original_topic = event.value.get(ORIGINAL_TOPIC_FIELD, "")
+        if not original_topic:
+            error = self.OriginalTopicError(event_label=event_label)
+            logging.critical(error, extra={"payload": event.value})
+            raise error
+
+        async with set_correlation_id(correlation_id):
+            await self._publisher.publish(
+                payload=event.value,
+                type_=type_,
+                topic=self._retry_topic,
+                key=event.key,
+            )
+            logging.info(
+                "Published an event with type '%s' to the retry topic '%s'",
+                type_,
+                self._retry_topic,
+            )
+
+    async def run(self, ignore: bool = False) -> None:
```

Comment:
Instead of passing a flag, we could pass a callback function that can inspect the event and then ignore or pass it - maybe even allow this callback to modify the payload when passing it to the retry topic.